engine/consensus/matching/engine.go

```diff
@@ -900,7 +935,158 @@ func (e *Engine) requestPending() error {
 		Int("finalized_blocks_without_result", len(missingBlocksOrderedByHeight)).
 		Msg("requesting receipts")
 	for _, blockID := range missingBlocksOrderedByHeight {
-		e.requester.EntityByID(blockID, filter.Any)
+		e.receiptRequester.EntityByID(blockID, filter.Any)
+	}
+
+	return nil
+}
+
+// requestPendingApprovals requests approvals for chunks that haven't collected
+// enough approvals. When the number of unsealed finalized blocks exceeds the
+// threshold, we go through the entire mempool of incorporated-results, which
+// haven't yet been sealed, and check which chunks need more approvals. We only
+// request approvals if the block incorporating the result is below the
+// threshold.
+//
+//                                   threshold
+//                              |                   |
+// ... <-- A <-- A+1 <- ... <-- D <-- D+1 <- ... -- F
+//       sealed       maxHeightForRequesting      final
+func (e *Engine) requestPendingApprovals() error {
+
+	// Skip requesting approvals if they are not required for sealing.
+	// TODO: this is only here temporarily to ease the migration to new chunk
+	// based sealing.
+	if !e.requireApprovals {
+		return nil
+	}
+
+	// last sealed block
+	sealed, err := e.state.Sealed().Head()
+	if err != nil {
+		return fmt.Errorf("could not get sealed height: %w", err)
+	}
+
+	// last finalized block
+	final, err := e.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("could not get finalized height: %w", err)
+	}
+
+	// only request if number of unsealed finalized blocks exceeds the threshold
+	log := e.log.With().
+		Uint64("finalized_height", final.Height).
+		Uint64("sealed_height", sealed.Height).
+		Uint64("approval_requests_threshold", e.approvalRequestsThreshold).
+		Logger()
+
+	maxHeightForRequesting := final.Height - e.approvalRequestsThreshold
+	if maxHeightForRequesting <= sealed.Height {
+		log.Debug().Msg("skip requesting approvals as number of unsealed finalized blocks is below threshold")
+		return nil
+	}
+
+	log.Info().Msg("requesting approvals")
```

Comment:
I would suggest to move this log to the end of the function with how many approvals were requested in total